dirty_cat/super_vectorizer.py

```diff
@@ -162,70 +165,86 @@ def __init__(self, *,
+        for column in X:
+            dtype = X[column].dtype
+            if issubclass(dtype.__class__, ExtensionDtype):
+                try:
+                    X[column] = X[column].astype(dtype.type)
+                except TypeError:
+                    pass
+        return X
+
+    def transform(self, X):
+        """Transform X separately by each transformer, concatenate results.
+
+        Parameters
+        ----------
+        X : {array-like, dataframe} of shape (n_samples, n_features)
+            The data to be transformed by subset.
 
-    def _transform(self, X) -> pd.DataFrame:
+        Returns
+        -------
+        X_t : {array-like, sparse matrix} of \
+                shape (n_samples, sum_n_components)
+            hstack of results of transformers. sum_n_components is the
+            sum of n_components (output dimension) over transformers. If
+            any result is a sparse matrix, everything will be converted to
+            sparse matrices.
+
+        """
         # Create a copy to avoid altering the original data.
         X = X.copy()
         # Convert to pandas DataFrame if not already.
         if not isinstance(X, pd.DataFrame):
             X = pd.DataFrame(X)
-
-        # Detect if the array contains missing values.
-        if _has_missing_values(X):
-            if self.handle_missing == '':
-                X = _replace_missing(X)
-            elif self.handle_missing == 'error':
-                raise ValueError('Array contains missing values')
-            else:
-                raise ValueError("Invalid 'handle_missing' value. "
-                                 "Expected any of {'', 'error'}, "
-                                 f"got {self.handle_missing}")
+            # Check the number of columns matches the fitted array's.
+            if X.shape[1] != len(self.columns_):
+                raise ValueError("Passed array does not match column count of "
+                                 f"fitter array. Got X.shape={X.shape}")
+            # If the DataFrame does not have named columns already,
+            # apply the learnt columns
+            if isinstance(X.columns, pd.RangeIndex):
+                X.columns = self.columns_
 
         if self.auto_cast:
-            from pandas.core.dtypes.base import ExtensionDtype
-            X = self._auto_cast_array(X)
-
-            if LooseVersion(sklearn.__version__) <= LooseVersion('0.22'):
-                # Cast pandas dtypes to numpy dtypes
-                # for earlier versions of sklearn
-                for column in X:
-                    dtype = X[column].dtype
-                    if issubclass(dtype.__class__, ExtensionDtype):
-                        try:
-                            X[column] = X[column].astype(dtype.type)
-                        except TypeError:
-                            pass
+            X.columns = self.columns_
+            X = X.astype(self.types_)
```

Comment:
Maybe if you move this line above, you won't be modifying the input dataframe, as this line assigns a new object to the dataframe. Hence, you will not need to copy the dataframe to address my comment above.

Reply:
Unfortunately it won't work in this context because we only use `X.astype` when `auto_cast` is `True`, but right after, we replace missing values regardless.   
We already perform a copy at the beginning as pointed out in a previous comment, so I'll leave it that way for now.